runtime/compiler/z/codegen/UncommonBCDCHKAddressNode.cpp

```diff
@@ -50,37 +50,61 @@ UncommonBCDCHKAddressNode::perform()
+               method == TR::com_ibm_dataaccess_DecimalData_convertLongToPackedDecimal_ ||
+               method == TR::com_ibm_dataaccess_PackedDecimal_addPackedDecimal_ ||
+               method == TR::com_ibm_dataaccess_PackedDecimal_subtractPackedDecimal_ ||
+               method == TR::com_ibm_dataaccess_PackedDecimal_multiplyPackedDecimal_ ||
+               method == TR::com_ibm_dataaccess_PackedDecimal_dividePackedDecimal_ ||
+               method == TR::com_ibm_dataaccess_PackedDecimal_remainderPackedDecimal_||
+               method == TR::com_ibm_dataaccess_PackedDecimal_shiftLeftPackedDecimal_ ||
+               method == TR::com_ibm_dataaccess_PackedDecimal_shiftRightPackedDecimal_) &&
+            node->getFirstChild()->getType().isBCD() &&
+            oldAddressNode->getDataType().isAddress() &&
+            oldAddressNode->getReferenceCount() > 1)
             {
-            TR::Node* pdOpNode = node->getFirstChild();
-            TR::Node* oldAddressNode = node->getSecondChild();
-            TR_ASSERT(pdOpNode && oldAddressNode, "Unexpected null PD opNode or address node under BCDCHK");
-            TR_ASSERT(oldAddressNode->getNumChildren() == 2, "Expecting 2 children under address node of BCDCHK.");
-
             TR::ILOpCodes addressOp = oldAddressNode->getOpCodeValue();
-            TR_ASSERT((addressOp == TR::aladd || addressOp == TR::aiadd), "Unexpected addressNode opcode");
-
-            if(oldAddressNode->getReferenceCount() > 1)
-               {
-               TR::Node* newAddressNode = TR::Node::create(node, addressOp, 2,
-                                                           oldAddressNode->getFirstChild(),
-                                                           oldAddressNode->getSecondChild());
-               node->setAndIncChild(1, newAddressNode);
-               oldAddressNode->decReferenceCount();
-               traceMsg(comp(), "%sReplacing node %s [%p] with [%p]\n",
-                        OPT_DETAILS,
-                        oldAddressNode->getOpCode().getName(),
-                        oldAddressNode,
-                        newAddressNode);
+            /* Expected tree structures for oldAddressNode
+             * 1. non off-heap mode
+             *     aladd/aiadd
+             *         load array_object
+             *         array_element_offset
+             *
+             * 2. off-heap enabled: loading first array element because offset is 0
+             *     aloadi <contiguousArrayDataAddrFieldSymbol>
+             *         load array_object
+             *
+             * 3. off-heap enabled
+             *     aladd/aiadd
+             *         aloadi <contiguousArrayDataAddrFieldSymbol>
+             *             load array_object
+             *         array_element_offset
+             */
+            TR_ASSERT_FATAL_WITH_NODE(node,
+                                      oldAddressNode->isDataAddrPointer() && oldAddressNode->getNumChildren() == 1
```

Comment:
Please add brackets here
```suggestion
                                      (oldAddressNode->isDataAddrPointer() && oldAddressNode->getNumChildren() == 1)
```